File: ch06/elixir/pbt/lib/checkout.ex
defmodule Checkout do
  @moduledoc false

  def total(item_list, price_list, specials) do
    if not valid_price_list(price_list),
      do: raise(RuntimeError, message: "invalid list of prices")

    if not valid_special_list(specials),
      do: raise(RuntimeError, message: "invalid list of specials")

    counts = count_seen(item_list)
    {counts_left, prices} = apply_specials(counts, specials)
    prices + apply_regular(counts_left, price_list)
  end

  def valid_price_list(prices) do
    unique =
      prices
      |> Enum.sort()
      |> Enum.dedup_by(&elem(&1, 0))

    length(prices) == length(unique)
  end

  defp valid_special_list([]), do: true

  defp valid_special_list(specials) do
    Enum.all?(specials, fn {_, x, _} -> x != 0 end)
  end

  defp count_seen(item_list) do
    count = &(&1 + 1)

    Enum.reduce(item_list, %{}, fn item, m ->
      Map.update(m, item, 1, count)
    end)
  end

  defp apply_specials(items, specials) do
    Enum.map_reduce(items, 0, fn {name, count}, price ->
      case List.keyfind(specials, name, 0) do
        nil ->
          {{name, count}, price}

        {_, needed, value} ->
          {{name, rem(count, needed)}, value * div(count, needed) + price}
      end
    end)
  end

  defp apply_regular(items, price_list) do
    Enum.sum(
      for {name, count} <- items do
        count * cost_of_item(name, price_list)
      end
    )
  end

  defp cost_of_item(name, price_list) do
    case List.keyfind(price_list, name, 0) do
      nil -> raise RuntimeError, message: "unknown item: #{name}"
      {_, price} -> price
    end
  end
end

File: ch07/erlang/pbt/src/checkout.erl
-module(checkout).
-export([valid_price_list/1, total/3]).

-type item() :: string().
-type price() :: integer().
-type special() :: {item(), pos_integer(), price()}.

-spec valid_price_list([{item(), price()}]) -> boolean().
valid_price_list(PriceList) ->
  length(PriceList) =:= length(lists:ukeysort(1, PriceList)).

-spec valid_special_list([special()]) -> boolean().
valid_special_list([]) ->
  true;
valid_special_list(Specials) ->
  lists:all(fun({_, X, _}) -> X =/= 0 end, Specials).

-spec total([item()], [{item(), price()}], [special()]) -> price().
total(ItemList, PriceList, Specials) ->
  valid_price_list(PriceList) orelse error(invalid_price_list),
  valid_special_list(Specials) orelse error(invalid_special_list),
  Counts = count_seen(ItemList),
  {CountsLeft, Prices} = apply_specials(Counts, Specials),
  Prices + apply_regular(CountsLeft, PriceList).

-spec count_seen([item()]) -> [{item(), pos_integer()}].
count_seen(ItemList) ->
  Count = fun(X) -> X + 1 end,
  maps:to_list(
    lists:foldl(fun(Item, M) -> maps:update_with(Item, Count, 1, M) end,
      maps:new(), ItemList)
  ).

-spec apply_specials([{item(), pos_integer()}], [special()]) ->
  {[{item(), pos_integer()}], price()}.
apply_specials(Items, Specials) ->
  lists:mapfoldl(fun({Name, Count}, Price) ->
      case lists:keyfind(Name, 1, Specials) of
        false ->
          {{Name, Count}, Price};
        {_, Needed, Value} ->
          {{Name, Count rem Needed},
            Value * (Count div Needed) + Price}
      end
  end, 0, Items).

-spec apply_regular([{item(), integer()}], [{item(), price()}]) -> price().
apply_regular(Items, PriceList) ->
  lists:sum([Count * cost_of_item(Name, PriceList) || {Name, Count} <- Items]).

cost_of_item(Name, PriceList) ->
  case proplists:get_value(Name, PriceList)  of
    undefined -> error({unknown_item, Name});
    Price -> Price
  end.